src/object/srv_obj.c

```diff
@@ -943,126 +1029,43 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
-				dti_cos_count, dti_cos, &obj_arg);
-		if (rc != 0) {
-			D_ERROR(DF_UOID": ds_obj_req_disp_prepare failed %d.\n",
-				DP_UOID(orw->orw_oid), rc);
-			D_GOTO(out, rc);
-		}
-		D_ASSERT(obj_arg != NULL);
-
-		rc = dss_ult_create(ds_obj_req_dispatch, obj_arg,
-				    DSS_ULT_IOFW, tag, 0, NULL);
-		if (rc != 0) {
-			D_ERROR(DF_UOID": ds_obj_req_dispatch failed %d.\n",
-				DP_UOID(orw->orw_oid), rc);
-			ds_obj_req_disp_arg_free(obj_arg);
-			D_GOTO(out, rc);
-		}
-
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_DISPATCH);
-	}
-
-	if (!resend) {
-		D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-
-		/* local RPC handler */
-		rc = ds_obj_rw_local_hdlr(rpc, tag, cont_hdl, cont,
-					  &ioh, true, map_ver, dth);
-		if (rc != 0)
-			D_ERROR(DF_UOID": rw_local_hdlr (update) failed %d.\n",
-				DP_UOID(orw->orw_oid), rc);
-
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_LOCAL);
-	}
-
-	/* wait dispatched IO's completion when needed */
-	if (dispatch) {
-		struct dtx_conflict_entry	*dces = NULL;
-
-		D_TIME_START(tls->ot_sp, OBJ_PF_UPDATE_WAIT);
-
-		rc1 = ds_obj_req_disp_wait(obj_arg, rc >= 0 ? &dces : NULL);
-		if (dces != NULL) {
-			/* XXX: The local modification has been done, but remote
-			 *	replica failed because of some uncommitted DTX,
-			 *	it may be caused by some garbage DTXs on remote
-			 *	replicas or leader has more information because
-			 *	of CoS cache. So handle (abort or commit) them
-			 *	firstly then retry.
-			 */
-
-			D_ASSERT(dth != NULL);
-
-			D_DEBUG(DB_TRACE, "Hit conflict DTX (%d)"DF_DTI" for "
-				DF_DTI", handle them and retry update.\n",
-				rc1, DP_DTI(&dces[0].dce_xid),
-				DP_DTI(&orw->orw_dti));
-
-			rc = dtx_conflict(cont->sc_hdl, dth,
-					  cont_hdl->sch_pool->spc_uuid,
-					  cont->sc_uuid, dces, rc1,
-					  cont_hdl->sch_pool->spc_map_version);
-			D_FREE(dces);
-			if (rc >= 0) {
-				resend = true;
-				flags |= ORF_RESEND;
-				dti_cos = dth->dth_dti_cos;
-				dti_cos_count = dth->dth_dti_cos_count;
-
-				D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_WAIT);
-
-				goto dispatch;
-			}
-		} else if (rc >= 0) {
-			rc = rc1;
-		}
-
-		D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE_WAIT);
-	}
-
+	exec_arg.rpc = rpc;
+	exec_arg.cont_hdl = cont_hdl;
+	exec_arg.cont = cont;
+again:
+	exec_arg.flags = flags;
+	/* Execute the operation on all targets */
+	rc = dtx_exec_ops(orw->orw_shard_tgts.ca_arrays,
+			  orw->orw_shard_tgts.ca_count, dth,
+			  ds_obj_tgt_update, &exec_arg);
 out:
-	if (dth != NULL)
-		rc = dtx_end(dth, cont_hdl, cont, rc);
+	/* Stop the distribute transaction */
+	rc = dtx_end(dth, cont_hdl, cont, rc);
+	if (rc == -DER_AGAIN) {
+		flags |= ORF_RESEND;
+		D_GOTO(again, rc);
+	}
 
 	ds_obj_rw_reply(rpc, rc, map_ver, &conflict);
-
 	D_TIME_END(tls->ot_sp, OBJ_PF_UPDATE);
 
-	if (cont_hdl) {
-		if (!cont_hdl->sch_cont)
-			ds_cont_child_put(cont); /* -1 for rebuild container */
+	if (cont_hdl)
 		ds_cont_hdl_put(cont_hdl);
-	}
+	if (cont)
+		ds_cont_child_put(cont);
```

Comment:
your changed code seems correct, then original code "if (!cont_hdl->sch_cont) ds_cont_child_put(cont)" was with bug of refcount leak?

Reply:
oh, the original implementation is not actually a bug. since it adds ds_cont_child_get() ds_check_container